packages/playwright-core/src/server/dom.ts

```diff
@@ -647,7 +647,12 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
         await Promise.all(localPaths.map(localPath => (
           fs.promises.access(localPath, fs.constants.F_OK)
         )));
+        const isDirectoryUpload = localPaths.length === 1 ? (await fs.promises.stat(localPaths[0])).isDirectory() : false;
```

Comment:
Perhaps only check this when `webkitdirectory` attribute is present?

Reply:
While file uploads to `input[webkitdirectory]` elements were possible before, this would mean we wait also for file uploads, while my change only waits for folder uploads. I prefer slightly only waiting for actual folder uploads, wdyt?

Reply:
I meant avoiding `fs.promises.stat` when `[webkitdirectory]` is not present. From what I see, we now always do this check unconditionally when passed a single file, which is the most common case.

Reply:
If you want to throw https://github.com/microsoft/playwright/pull/31165#discussion_r1630015449 from both sides, this means we'd need to call stat from the client and server. So there is no way around that anyways?